firebase-firestore/src/main/java/com/google/firebase/firestore/model/Document.java

```diff
@@ -36,59 +40,107 @@
   public static Comparator<Document> keyComparator() {
     return KEY_COMPARATOR;
   }
 
-  private final ObjectValue data;
-
-  private final DocumentState documentState;
+  /** Creates a new Document from an existing ObjectValue. */
+  public static Document fromObjectValue(
+      DocumentKey key, SnapshotVersion version, ObjectValue data, DocumentState documentState) {
+    return new Document(key, version, documentState) {
+      @Nullable
+      @Override
+      public com.google.firestore.v1.Document getProto() {
+        return null;
+      }
+
+      @Override
+      public ObjectValue getData() {
+        return data;
+      }
+
+      @Nullable
+      @Override
+      public FieldValue getField(FieldPath path) {
+        return data.get(path);
+      }
+    };
+  }
 
   /**
-   * Memoized serialized form of the document for optimization purposes (avoids repeated
-   * serialization). Might be null.
+   * Creates a new Document from a Proto representation.
+   *
+   * <p>The Proto is only converted to an ObjectValue if the consumer calls `getData()`.
    */
-  private final com.google.firestore.v1.Document proto;
-
-  public @Nullable com.google.firestore.v1.Document getProto() {
-    return proto;
+  public static Document fromProto(
+      RemoteSerializer serializer,
+      com.google.firestore.v1.Document proto,
+      DocumentState documentState) {
+    DocumentKey key = serializer.decodeKey(proto.getName());
+    SnapshotVersion version = serializer.decodeVersion(proto.getUpdateTime());
+
+    hardAssert(
+        !version.equals(SnapshotVersion.NONE), "Found a document Proto with no snapshot version");
+
+    return new Document(key, version, documentState) {
+      private ObjectValue memoizedData = null;
+
+      @Override
+      public com.google.firestore.v1.Document getProto() {
+        return proto;
+      }
+
+      @Override
+      public ObjectValue getData() {
+        if (memoizedData != null) {
+          return memoizedData;
+        } else {
+          memoizedData = serializer.decodeFields(proto.getFieldsMap());
+          return memoizedData;
+        }
+      }
+
+      @Nullable
+      @Override
+      public FieldValue getField(FieldPath path) {
+        if (memoizedData != null) {
+          return memoizedData.get(path);
+        } else {
+          // Instead of deserializing the full Document proto, we only deserialize the value at
+          // the requested field path. This speeds up Query execution as query filters can discard
+          // documents based on a single field.
+          Value value = proto.getFieldsMap().get(path.getFirstSegment());
+          for (int i = 1; value != null && i < path.length(); ++i) {
+            if (value.getValueTypeCase() != Value.ValueTypeCase.MAP_VALUE) {
+              return null;
+            }
+            value = value.getMapValue().getFieldsMap().get(path.getSegment(i));
+          }
+          return value != null ? serializer.decodeValue(value) : null;
```

Comment:
Our `OrderBy` implementation does not implement a Schwartzian transform so we'll be performing these decodes for each comparison. For a large collection this could potentially eat into the gains you're observing here. I wonder if a cache of `FieldPath` to `FieldValue` would hurt here?

Reply:
I added a cache - but this has to be a ConcurrentHashMap since this function can be called from user land. Please do shout if that is not actually an issue :)

Reply:
I think the impact of this caching will be hard to quantify. I thought about adding this initially, but was only won over by your comments earlier. We do re-run getField() excessively during the computation of our Views.

I changed the code to lazy-init the field value cache. I do want to change the comparator to compare on the Protos directly, at which point we can remove the cache again.